text/0000-placement-by-return.md

```diff
@@ -0,0 +1,1097 @@
+
+```rust
+// This function will allocate its return value directly in the return slot.
+fn good() -> [i32; 1_000_000] {
+    [1; 1_000_000]
+}
+// This function will return a raw slice, and can also be called as `Box::new_with(good_dst)`
+fn good_dst() -> [i32] {
+    let n = 1_000_000;
+    [1; n]
+}
+// This function will copy the array when it returns.
+fn bad() -> [i32; 1_000_000] {
+    let mut arr = [0; 1_000_000];
+    for i in 0..1_000_000 {
+        arr[i] = i;
+    }
+    arr
+}
+// This function will compile successfully with #![feature(unsized_locals)]
+// but it will allocate the array on the stack, which is probably very bad.
+fn bad_dst() {
+    fn takes_dst(_k: [i32]) {}
+    fn returns_dst() -> [i32] { let n = 1_000_000; [1; n] }
+    takes_dst(returns_dst())
+}
+```
+
+This is guaranteed to see through if expressions, function calls, literals, unsafe blocks, and the return statement. It is not guaranteed to see through variable assignment or break with value.
+
+```rust
+// "good" functions will write their results directly in the caller-provided slot
+fn good() -> Struct {
+    if something() { Struct { ... } } else { Struct { ... } }
+}
+fn good() -> Struct {
+    loop { return Struct { ... } }
+}
+fn good() -> Struct2 {
+    Struct2 { member: Struct { ... } }
+}
+// "bad" functions will not necessarily do that
+fn bad() -> Struct {
+    loop { break Struct { ... } }
+}
+fn bad() -> Struct {
+    let q = Struct { ... };
+    q
+}
+fn bad() -> Struct2 {
+    let q = Struct { ... }
+    Struct2 { member: q }
+}
+```
+
+In other words, Rust does not currently guarantee Named Return Value Optimization.
+
+In the rustonomicon, it should mention that the `write_return_with` intrinsic can be used to build a function that's equivalent to `bad()`:
+
+```rust
+use std::mem::{write_return_with, ReturnSlot};
+use std::alloc::Layout;
+fn not_bad() -> [i32] {
+    unsafe {
+        write_return_with(Layout::new::<[i32; 1_000_000]>(), |arr: *mut u8| {
+            let arr: &mut [i32] = slice::from_raw_parts_mut(arr, 1_000_000);
+            for i in 0..1_000_000 {
+                arr[i] = i;
+            }
+            arr
+        })
+    }
+}
+```
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+When a function returns a value while respecting the constraints described above, the value's observable address will not change, if the value's type is:
+
+- Larger than an implementation-defined size (eg a register).
+- Unsized.
+
+(note: we could add a trait in a future RFC that would guarantee a type is always emplaced, even when it's register-sized, if use cases appear that need the "unchanged address" guarantee)
+
+## Did you say I can return unsized types?
+
+A function that directly returns an unsized type should be compiled into two functions, essentially as a special kind of generator:
```

Comment:
I feel like this could use a bit of clarification on top of the examples. Basically clarify that you're splitting the function into the half that decides what layout the data will take in memory, and the half that will actually write to the allocated memory (that being on either the stack or the heap depending on the caller).